platform/consul/controller.go

```diff
@@ -0,0 +1,219 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package consul
+
+import (
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+	"github.com/hashicorp/consul/api"
+	"istio.io/pilot/model"
+)
+
+// Controller communicates with Consul and monitors for changes
+type Controller struct {
+	client     *api.Client
+	dataCenter string
+	monitor    Monitor
+}
+
+// NewController creates a new Consul controller
+func NewController(addr, datacenter string) (*Controller, error) {
+	conf := api.DefaultConfig()
+	conf.Address = addr
+
+	client, err := api.NewClient(conf)
+	return &Controller{
+		monitor:    NewConsulMonitor(client, time.Second*3),
+		client:     client,
+		dataCenter: datacenter,
+	}, err
+}
+
+// Services list declarations of all services in the system
+func (c *Controller) Services() []*model.Service {
+	data := c.getServices()
+
+	services := make([]*model.Service, 0, len(data))
+	for name := range data {
+		endpoints := c.getCatalogService(name, nil)
+		services = append(services, convertService(endpoints))
+	}
+
+	return services
+}
+
+// GetService retrieves a service by host name if it exists
+func (c *Controller) GetService(hostname string) (*model.Service, bool) {
+	// Get actual service by name
+	name, err := parseHostname(hostname)
+	if err != nil {
+		glog.V(2).Infof("parseHostname(%s) => error %v", hostname, err)
+		return nil, false
+	}
+
+	endpoints := c.getCatalogService(name, nil)
+	if len(endpoints) == 0 {
+		return nil, false
+	}
+
+	out := convertService(endpoints)
+
+	return out, true
+}
+
+func (c *Controller) getServices() map[string][]string {
+	data, _, err := c.client.Catalog().Services(nil)
+	if err != nil {
+		glog.Warningf("Could not retrieve services from consul: %v", err)
+		return make(map[string][]string)
+	}
+
+	return data
+}
+
+func (c *Controller) getCatalogService(name string, q *api.QueryOptions) []*api.CatalogService {
+	endpoints, _, err := c.client.Catalog().Service(name, "", q)
+	if err != nil {
+		glog.Warningf("Could not retrieve service catalogue from consul: %v", err)
+		return []*api.CatalogService{}
+	}
+
+	return endpoints
+}
+
+// Instances retrieves instances for a service and its ports that match
+// any of the supplied tags. All instances match an empty tag list.
+func (c *Controller) Instances(hostname string, ports []string, tags model.TagsList) []*model.ServiceInstance {
+	// Get actual service by name
+	name, err := parseHostname(hostname)
+	if err != nil {
+		glog.V(2).Infof("parseHostname(%s) => error %v", hostname, err)
+		return nil
+	}
+
+	endpoints := c.getCatalogService(name, nil)
+
+	instances := []*model.ServiceInstance{}
+	for _, endpoint := range endpoints {
+		inst := convertInstance(endpoint)
+		if tags.HasSubsetOf(inst.Tags) && portMatch(inst, ports) {
```

Comment:
`portMatch` could be more efficient. check against a `map[string]bool` populated with the port names.